### PR TITLE
feat: add initial sources

### PR DESCRIPTION
**Pull request by utkuozdemir**
The initial implementation.

Depends on: https://github.com/siderolabs/omni/pull/449

### CODE REVIEW

File: .kres.yaml [hunk 1]
```diff
@@ -2,6 +2,7 @@
 kind: common.Image
 name: image-omni-cloud-provider-qemu
 spec:
+  baseImage: ghcr.io/siderolabs/talosctl:v1.7.5
```
Comment (utkuozdemir) [inline .kres.yaml]: This provisioner needs `talosctl`, as the provisioning code imported from Talos requires it to spawn more processes over it. The `talosctl` binary is searched in the working directory and in the PATH by this provisioner. When we use it as the base image, it is found in the current working directory.

File: api/specs/specs.proto [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+message QemuMachineSpec {
+  string uuid = 1;
+  string disk_path = 2;
+}
+
+message QemuMachineAllocationSpec {
+  string talos_version = 1;
+  string schematic_id = 2;
+}
```
Comment (utkuozdemir) [inline api/specs/specs.proto:15]: These are the resources internal to this provider, living in the `cloud-provider:qemu` namespace of Omni's state.

File: cmd/omni-cloud-provider-qemu/main.go [hunk 3]
```diff
@@ -5,4 +5,114 @@
...
+		"the endpoint of the Omni API, if not set, defaults to OMNI_ENDPOINT env var.")
+	rootCmd.Flags().StringVar(&meta.ProviderID, "id", meta.ProviderID, "the id of the cloud provider, it is used to match the resources with the cloud provider label.")
+	rootCmd.Flags().StringVar(&rootCmdArgs.talosctlPath, "talosctl-path", "", "the path to the talosctl binary, when unset, "+
+		"it is searched in the current working directory and in the $PATH.")
+	rootCmd.Flags().StringVar(&rootCmdArgs.subnetCIDR, "subnet-cidr", "172.42.0.0/24", "the CIDR of the subnet to use for the QEMU VMs.")
+	rootCmd.Flags().StringSliceVar(&rootCmdArgs.nameservers, "nameservers", []string{"1.1.1.1", "1.0.0.1"}, "the nameservers to use for the QEMU VMs.")
+	rootCmd.Flags().StringVar(&rootCmdArgs.imageFactoryPXEURL, "image-factory-pxe-url", "https://pxe.factory.talos.dev", "the URL of the image factory PXE server.")
+	rootCmd.Flags().IntVar(&rootCmdArgs.ipxeServerPort, "ipxe-server-port", 42420, "the port of the iPXE server.")
+	rootCmd.Flags().IntVar(&rootCmdArgs.numMachines, "num-machines", 8, "the number of machines to provision.")
```
Comment (utkuozdemir) [inline cmd/omni-cloud-provider-qemu/main.go:118]: This is the total pool of machines the provider provisions on startup. They go into an ipxe boot loop until they are assigned to a `MachineRequest`.

File: .kres.yaml (hunk 1 above)
Comment (frezbo) [reply .kres.yaml]: i think we can just use a copy stage and copy over from the talosctl image

File: hack/release.toml [hunk 4]
```diff
@@ -1,9 +1,9 @@
 # commit to be tagged for the new release
 commit = "HEAD"
 
-project_name = "omni-cloud-provider-qemu"
-github_repo = "siderolabs/omni-cloud-provider-qemu"
-match_deps = "^github.com/(siderolabs/[a-zA-Z0-9-]+)$"
+project_name = ""
```
Comment (frezbo) [inline hack/release.toml]: weird, this shoudn't be changed

File: internal/provisioner/provisioner.go [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to create provisioner: %w", err)
+	}
+
+	gatewayAddr, err := sideronet.NthIPInNetwork(provisioner.subnetCIDR, 1)
+	if err != nil {
+		return fmt.Errorf("failed to get gateway address: %w", err)
+	}
+
+	cniBundleURL := "https://github.com/siderolabs/talos/releases/download/v1.7.5/talosctl-cni-bundle-amd64.tar.gz"
```
Comment (frezbo) [inline internal/provisioner/provisioner.go]: probably we can pick latest always?

File: .kres.yaml (hunk 1 above)
Comment (utkuozdemir) [reply .kres.yaml]: I look into that, but it looked like I can copy over from other images (in the same dockerfile), but not from an arbitrary image. Might be wrong though.

Also, does this have any downsides? There is nothing in the talosctl base image other than talosctl.

File: hack/release.toml (hunk 4 above)
Comment (utkuozdemir) [reply hack/release.toml]: reverted that manually, they stay after rekres.

File: internal/provider/provider.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+	return eg.Wait()
+}
+
+// watchSysVersion serves as a health check - if the connection to the Omni API is lost (e.g., due to an Omni restart),
+// this will return an error and the provider will crash.
+func (provider *Provider) watchSysVersion(ctx context.Context, st state.State) error {
+	eventCh := make(chan state.Event)
+
+	if err := st.Watch(ctx, system.NewSysVersion(omniresources.EphemeralNamespace, system.SysVersionID).Metadata(), eventCh); err != nil {
```
Comment (Unix4ever) [inline internal/provider/provider.go:156]: I wonder if we should add a proper health endpoint.

**Review by Unix4ever — APPROVED** [review]

File: internal/provisioner/provisioner.go (hunk 5 above)
Comment (utkuozdemir) [reply internal/provisioner/provisioner.go]: good point, done 👍

File: internal/provider/provider.go (hunk 6 above)
Comment (utkuozdemir) [reply internal/provider/provider.go:156]: Yep, that's a good idea. Here the problem I work around is, when Omni gets restarted, the `state` I have over the client API gets broken, but the runtime doesn't become aware of it. I couldn't find a way to make it aware of it (using GRPC options etc.). Maybe there is one, but I couldn't find any.

So not sure if it's an Omni problem - maybe a COSI one?

**Comment by utkuozdemir** [comment]
Holding this until the Omni side gets merged

**Pull request merged**